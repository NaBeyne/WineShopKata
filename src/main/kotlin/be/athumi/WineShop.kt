package be.athumi

class WineShop(var items: List<Wine>) {
    fun updateWines() {
        val wineWrapperFactory = WineWrapperFactory()
        for (wine in items) {
            wineWrapperFactory.getWrappedWine(wine).updateWine()
        }
    }
}