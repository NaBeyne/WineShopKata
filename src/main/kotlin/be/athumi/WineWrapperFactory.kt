package be.athumi

import be.athumi.wines.*

private const val EVENT_IDENTIFIER = "Event"
private const val CONSERVATO_IDENTIFIER = "Conservato"
private const val WINE_BREWED_BY_ALEXANDER_THE_GREAT = "Wine brewed by Alexander the Great"

class WineWrapperFactory {

    fun getWrappedWine(wine: Wine) : WineWrapper {
        return when {
            wine.name.contains(CONSERVATO_IDENTIFIER) -> ConservatoWineWrapper(wine)
            wine.name.startsWith(EVENT_IDENTIFIER) -> EventWineWrapper(wine)
            wine.name == WINE_BREWED_BY_ALEXANDER_THE_GREAT -> WineBrewedByAlexanderTheGreatWrapper(wine)
            else -> RegularWineWrapper(wine)
        }
    }

}