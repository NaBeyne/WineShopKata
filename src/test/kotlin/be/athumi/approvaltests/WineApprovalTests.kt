package be.athumi.approvaltests

import be.athumi.Wine
import be.athumi.WineShop
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test

class WineApprovalTests {

    private val wines = listOf(
        Wine("Standard Wine", 20, 10),
        Wine("Another Standard Wine", 7, 5),
        Wine("Negative Standard Wine", -5, 5),
        Wine("Bourdeaux Conservato", 0, 2),
        Wine("Bourgogne Conservato", 98, 2),
        Wine("Wine brewed by Alexander the Great", 150, 0),
        Wine("Wine brewed by Alexander the Great", 80, -1),
        Wine("Wine brewed by Alexander the Great", -1, 10),
        Wine("Event Wine", 20, 15),
        Wine("Event Wine", 49, 10),
        Wine("Event Wine", 99, 2),
        Wine("Eco Brilliant Wine", 6, 3)
    )

    @Test
    fun updateWineShopFor1Years() {
        updateWineShopForAmountOfYears(1)
    }

    @Test
    fun updateWineShopFor2Years() {
        updateWineShopForAmountOfYears(2)
    }

    @Test
    fun updateWineShopFor3Years() {
        updateWineShopForAmountOfYears(3)
    }

    @Test
    fun updateWineShopFor5Years() {
        updateWineShopForAmountOfYears(5)
    }

    @Test
    fun updateWineShopFor10Years() {
        updateWineShopForAmountOfYears(10)
    }

    @Test
    fun updateWineShopFor20Years() {
        updateWineShopForAmountOfYears(20)
    }

    @Test
    fun updateWineShopFor30Years() {
        updateWineShopForAmountOfYears(30)
    }

    @Test
    fun updateWineShopFor50Years() {
        updateWineShopForAmountOfYears(50)
    }

    private fun updateWineShopForAmountOfYears(amountOfYears: Int) {
        val wineShop = WineShop(wines)
        Approvals.verifyAll("Years", (0..amountOfYears)) { year ->
            """-------- year $year --------${'\n'}${wines.joinToString("\n")}""".trimIndent()
                .also {  wineShop.updateWines() }
        }
    }
}