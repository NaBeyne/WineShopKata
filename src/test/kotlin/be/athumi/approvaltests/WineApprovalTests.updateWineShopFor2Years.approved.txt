Years


-------- year 0 --------
Wine(name=Standard Wine, price=20, expiresInYears=10)
Wine(name=Another Standard Wine, price=7, expiresInYears=5)
Wine(name=Negative Standard Wine, price=-5, expiresInYears=5)
Wine(name=Bourdeaux Conservato, price=0, expiresInYears=2)
Wine(name=Bourgogne Conservato, price=98, expiresInYears=2)
Wine(name=Wine brewed by Alexander the Great, price=150, expiresInYears=0)
Wine(name=Wine brewed by Alexander the Great, price=80, expiresInYears=-1)
Wine(name=Wine brewed by Alexander the Great, price=-1, expiresInYears=10)
Wine(name=Event Wine, price=20, expiresInYears=15)
Wine(name=Event Wine, price=49, expiresInYears=10)
Wine(name=Event Wine, price=99, expiresInYears=2)
Wine(name=Eco Brilliant Wine, price=6, expiresInYears=3)
-------- year 1 --------
Wine(name=Standard Wine, price=19, expiresInYears=9)
Wine(name=Another Standard Wine, price=6, expiresInYears=4)
Wine(name=Negative Standard Wine, price=0, expiresInYears=4)
Wine(name=Bourdeaux Conservato, price=1, expiresInYears=1)
Wine(name=Bourgogne Conservato, price=99, expiresInYears=1)
Wine(name=Wine brewed by Alexander the Great, price=150, expiresInYears=0)
Wine(name=Wine brewed by Alexander the Great, price=80, expiresInYears=-1)
Wine(name=Wine brewed by Alexander the Great, price=0, expiresInYears=10)
Wine(name=Event Wine, price=21, expiresInYears=14)
Wine(name=Event Wine, price=50, expiresInYears=9)
Wine(name=Event Wine, price=100, expiresInYears=1)
Wine(name=Eco Brilliant Wine, price=4, expiresInYears=2)
-------- year 2 --------
Wine(name=Standard Wine, price=18, expiresInYears=8)
Wine(name=Another Standard Wine, price=5, expiresInYears=3)
Wine(name=Negative Standard Wine, price=0, expiresInYears=3)
Wine(name=Bourdeaux Conservato, price=2, expiresInYears=0)
Wine(name=Bourgogne Conservato, price=100, expiresInYears=0)
Wine(name=Wine brewed by Alexander the Great, price=150, expiresInYears=0)
Wine(name=Wine brewed by Alexander the Great, price=80, expiresInYears=-1)
Wine(name=Wine brewed by Alexander the Great, price=0, expiresInYears=10)
Wine(name=Event Wine, price=22, expiresInYears=13)
Wine(name=Event Wine, price=51, expiresInYears=8)
Wine(name=Event Wine, price=100, expiresInYears=0)
Wine(name=Eco Brilliant Wine, price=2, expiresInYears=1)
